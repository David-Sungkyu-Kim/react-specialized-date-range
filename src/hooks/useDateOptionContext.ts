import * as React from "react";
import {
  DatePickerOptionActionsContext,
  DatePickerOptionValueContext,
} from "../context/DatePickerOptionProvider";

function useDatePickerOptionValuesContext() {
  const context = React.useContext(DatePickerOptionValueContext);
  if (context === undefined)
    throw new Error("useDatePickerOptionValuesContext should be used within DateContext.Provider");

  return context;
}

function useDatePickerOptionActionsContext() {
  const context = React.useContext(DatePickerOptionActionsContext);
  if (context === undefined)
    throw new Error("useDatePickerOptionActionsContext should be used within DateContext.Provider");

  return context;
}

export { useDatePickerOptionValuesContext, useDatePickerOptionActionsContext };
