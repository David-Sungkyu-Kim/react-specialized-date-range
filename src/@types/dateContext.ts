import { IDate } from "./date";

export interface IDateContextValues {
  startDate: IDate;
  endDate: IDate;
  single: IDate;
  [key: string]: IDate;
}

export interface IDateContextActions {
  changeBiggerUnit: (standard: string) => void;
  changeTitle: (standard: string, arrow: string) => void;
  changeHighlightDate: (standard: string, dateStr: string, format: string, type: string) => void;
  changeMonth: (standard: string, index: number) => void;
  changeYear: (standard: string, year: number) => void;
  changeDecade: (standard: string, decade: number) => void;
  setSelectedDate: (double: boolean, value: string[], format: string) => void;
  setToDisabledEndDate: (double: boolean, disabledEndDate: string) => void;
}

export interface IDateContext {
  value: IDateContextValues;
  action: IDateContextActions;
}

export interface IDatePickerContextValues {
  width: string;
  height: string;
  double: boolean;
  startDayOfWeek: string;
  disabledDates: string[] | null;
  mode: string;
  placement: string;
  format: string;
  locale?: string;
  placeholder: string;
  readOnly?: boolean;
  value: string[];
  formatSeparator: () => string;
  onChange: (e: any) => void;
}

export interface IDatePickerContextActions {
  setInitOption: (option: IDatePickerContextValues) => void;
}
