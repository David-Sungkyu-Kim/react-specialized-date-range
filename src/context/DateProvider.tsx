import * as React from "react";
import {
  onChangeTitle,
  onChangeBiggerUnit,
  onChangeMonth,
  onChangeYearOrDecade,
  onChangeDate,
  onSetSelectDate,
  onSetToDisabledEndDate,
} from "../utils/date";
import { IDateContext, IDateContextActions, IDateContextValues } from "../@types/dateContext";
import {
  initialDate,
  initialDateAction,
  initialDateState,
} from "../utils/constants/initialContext";

export const DateValuesContext = React.createContext<IDateContextValues>(initialDateState);
export const DateActionsContext = React.createContext<IDateContextActions>(initialDateAction);

export const DateContext = React.createContext<IDateContext>(initialDate);

const DateProvider = ({ children }: { children: React.ReactNode }) => {
  const [date, setDate] = React.useState<IDateContextValues>(initialDateState);

  const actions = React.useMemo(
    () => ({
      changeBiggerUnit(standard: string) {
        setDate((prev) => ({
          ...prev,
          [standard]: onChangeBiggerUnit(prev[standard]),
        }));
      },
      changeTitle(standard: string, arrow: string) {
        setDate((prev) => ({
          ...prev,
          [standard]: onChangeTitle(prev[standard], arrow),
        }));
      },
      changeHighlightDate(standard: string, dateStr: string, format: string, type: string) {
        setDate((prev) => onChangeDate(prev, standard, dateStr, format, type));
      },
      changeMonth(standard: string, index: number) {
        setDate((prev) => ({
          ...prev,
          [standard]: onChangeMonth(prev[standard], index),
        }));
      },
      changeYear(standard: string, year: number) {
        setDate((prev) => ({
          ...prev,
          [standard]: onChangeYearOrDecade(prev[standard], year),
        }));
      },
      changeDecade(standard: string, decade: number) {
        setDate((prev) => ({
          ...prev,
          [standard]: onChangeYearOrDecade(prev[standard], decade),
        }));
      },
      setSelectedDate(double: boolean, value: string[], format: string) {
        setDate((prev) => (value ? onSetSelectDate(prev, double, value, format) : prev));
      },
      setToDisabledEndDate(double: boolean, disabledEndDate: string) {
        setDate((prev) => onSetToDisabledEndDate(prev, double, disabledEndDate));
      },
    }),
    []
  );

  // return (
  //   <DateActionsContext.Provider value={actions}>
  //     <DateValuesContext.Provider value={date}>{children}</DateValuesContext.Provider>
  //   </DateActionsContext.Provider>
  // );
  return (
    <DateContext.Provider value={{ value: date, action: actions }}>{children}</DateContext.Provider>
  );
};

export default DateProvider;
