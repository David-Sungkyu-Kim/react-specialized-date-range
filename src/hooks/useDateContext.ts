import * as React from "react";
import { DateActionsContext, DateContext, DateValuesContext } from "../context/DateProvider";

function useDateValuesContext() {
  const context = React.useContext(DateValuesContext);
  if (context === undefined)
    throw new Error("useDateValuesContext should be used within DateContext.Provider");

  return context;
}

function useDateActionsContext() {
  const context = React.useContext(DateActionsContext);
  if (context === undefined)
    throw new Error("useDateActionsContext should be used within DateContext.Provider");

  return context;
}

function useDateContext() {
  const context = React.useContext(DateContext);

  if (context === undefined)
    throw new Error("useDateContext should be used within DateContext.Provider");

  return context;
}

export { useDateValuesContext, useDateActionsContext, useDateContext };
