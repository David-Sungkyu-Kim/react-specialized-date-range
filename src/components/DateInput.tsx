import * as React from "react";
import { useDatePickerOptionValuesContext } from "../hooks/useDateOptionContext";
import { useDateContext } from "../hooks/useDateContext";
import { useTranslation } from "react-i18next";

import {
  checkSetFormatRegExr,
  convertToDeafultFormat,
  formattingNumToDate,
} from "../utils/dateFormat";
import { EMode, EStandard, EType, IDateInputProps } from "../@types/date";
import { IDatePickerContextValues } from "../@types/dateContext";

export default function DateInput({ standard, setIsActive, value }: IDateInputProps) {
  const { t } = useTranslation();
  const { value: date, action } = useDateContext();
  const option: IDatePickerContextValues = useDatePickerOptionValuesContext();
  const { disabledDates, placeholder, mode, format } = option;

  const [text, setText] = React.useState("");

  React.useEffect(() => {
    value && setText(value);
  }, []);

  React.useEffect(() => {
    // NOTE: Error 관리 - format 구분자와 disabledDates의 배열 원소들의 구분자가 달랐을 때 나오는 error
    if (disabledDates) {
      const disabledDatesStart = checkSetFormatRegExr(format, disabledDates[0]);
      const disabledDatesEnd = checkSetFormatRegExr(format, disabledDates[1]);

      if (!disabledDatesStart || !disabledDatesEnd)
        throw new Error("Set limits according to the format you set.");
    }
  }, [disabledDates]);

  React.useEffect(() => {
    date[standard] &&
      (date[standard].selectedDate || !value) &&
      setText(date[standard].selectedDate);
  }, [date[standard] && date[standard].selectedDate]);

  const handleChangeDate = React.useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = event.currentTarget;

      if (value.length <= 10) {
        const newDate = formattingNumToDate(value, format);
        const writeDay = convertToDeafultFormat(newDate, format);
        const disabledDateStart = disabledDates
          ? convertToDeafultFormat(disabledDates[0], format)
          : "";
        const disabledDateEnd = disabledDates
          ? convertToDeafultFormat(disabledDates[1], format)
          : "";

        setText(newDate);

        if (newDate.length === 10 || !newDate) {
          // NOTE: 설정한 날짜 범위가 아닐 경우 제일 마지막으로 설정했던 값으로 변한다. <초기화 시킬 수 있음>

          if (!disabledDates || !newDate)
            return action.changeHighlightDate(standard, newDate, format, EType.INPUT);

          disabledDateStart < writeDay && (writeDay < disabledDateEnd || !disabledDateEnd)
            ? action.changeHighlightDate(standard, newDate, format, EType.INPUT)
            : action.changeHighlightDate(standard, "", format, EType.INPUT);
          // : action.changeHighlightDate(standard, date[standard].selectedDate, format, EType.INPUT);
        }
      }
    },
    [text]
  );

  return (
    <div className='inputWrapper'>
      {standard !== EStandard.SINGLE && (
        <label className='inputLabel' role='label'>
          {standard === EStandard.STARTDATE
            ? `${t(EStandard.STARTDATE)}:`
            : `${t(EStandard.ENDDATE)}:`}
        </label>
      )}

      <div className='inputBasicWrapper'>
        <input
          role='input'
          className='inputValue'
          type='text'
          placeholder={placeholder}
          value={text}
          onChange={(e) => handleChangeDate(e)}
          onClick={() => mode === EMode.BASIC && setIsActive(true)}
        />
      </div>
    </div>
  );
}
