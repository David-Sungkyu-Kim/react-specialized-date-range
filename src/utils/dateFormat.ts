import { ELanguage, EUnit } from "../@types/date";
import i18n from "../lang/i18n";
import { months, thisDay, thisMonth, thisYear } from "./constants/date";

export const getWeekday = (year: number, month: number, day: number, locale: string) => {
  return new Date(year, month, day).toLocaleString(locale, {
    weekday: "short",
  });
};

export const translateOneToTenFormat = (date: number): string => {
  return date >= 10 ? String(date) : `0${date}`;
};

export const todayDashFormat = (year = thisYear, month = thisMonth, date = thisDay) => {
  return `${year}-${translateOneToTenFormat(month)}-${translateOneToTenFormat(date)}`;
};

export const convertTitleToUnit = (unit: string, year: number, month: number) => {
  if (unit === EUnit.DECADE) return `${(year + "").slice(0, 2)}00-${(year + "").slice(0, 2)}90`;
  if (unit === EUnit.YEAR) return `${(year + "").slice(0, 3)}0-${(year + "").slice(0, 3)}9`;
  if (unit === EUnit.MONTH) return `${year}`;
  if (unit === EUnit.DAY) {
    if (i18n.language === ELanguage.KO) return translateTitleToKo(year, month);
    if (i18n.language === ELanguage.EN) return translateTitleToEn(year, month);
    if (i18n.language === ELanguage.JA) return translateTitleToJa(year, month);
  }
};

const translateTitleToKo = (year: number, month: number) => `${year} ${month}월`;
const translateTitleToEn = (year: number, month: number) => `${months[month - 1]} ${year}`;
const translateTitleToJa = (year: number, month: number) => `${year}年 ${month}月`;
