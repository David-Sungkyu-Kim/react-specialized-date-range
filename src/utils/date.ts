import { EStandard, EUnit, IDate } from "../@types/date";
import { IDateContextValues } from "../@types/dateContext";
import { translateOneToTenFormat } from "./dateFormat";
import { refer } from "./dateOption";
import { changeCentury, changeDecade, changeMonth, changeYear } from "./dateTitle";

export const onChangeTitle = (prevDate: IDate, arrow: string) => {
  if (prevDate.unit === EUnit.DAY) return changeMonth(prevDate, arrow);
  if (prevDate.unit === EUnit.MONTH) return changeYear(prevDate, arrow);
  if (prevDate.unit === EUnit.YEAR) return changeDecade(prevDate, arrow);
  if (prevDate.unit === EUnit.DECADE) return changeCentury(prevDate, arrow);
};

export const onChangeBiggerUnit = (prevDate: IDate) => {
  if (prevDate.unit === EUnit.DECADE) return prevDate;

  let unit = prevDate.unit;

  if (prevDate.unit === EUnit.DAY) unit = EUnit.MONTH;
  if (prevDate.unit === EUnit.MONTH) unit = EUnit.YEAR;
  if (prevDate.unit === EUnit.YEAR) unit = EUnit.DECADE;

  return { ...prevDate, unit };
};

export const onChangeDateByCalendar = (
  prevDate: IDateContextValues,
  standard: string,
  clickedDate: any
) => {
  const { selectedDate } = prevDate[refer(standard)];

  const clickedDateArr = clickedDate.date.split("-");
  let selectedYear = Number(clickedDateArr[0]);
  let selectedMonth = Number(clickedDateArr[1]);
  const selectedDay = clickedDateArr[2];

  if (selectedMonth) {
    if (selectedMonth === 13) {
      selectedYear++;
      selectedMonth = 1;
    }
  } else {
    selectedYear--;
    selectedMonth = 12;
  }

  let newSelectedDate = {
    year: selectedYear,
    month: selectedMonth,
    selectedDate: `${selectedYear}-${translateOneToTenFormat(selectedMonth)}-${selectedDay}`,
  };

  if (
    standard !== EStandard.SINGLE &&
    ((refer(standard) === EStandard.ENDDATE && clickedDate.date > selectedDate) ||
      (refer(standard) === EStandard.STARTDATE && clickedDate.date < selectedDate))
  ) {
    return {
      ...prevDate,
      [standard]: { ...prevDate[standard], ...newSelectedDate },
      [refer(standard)]: { ...prevDate[refer(standard)], selectedDate: "" },
    };
  }

  return { ...prevDate, [standard]: { ...prevDate[standard], ...newSelectedDate } };
};

export const onChangeDateByInput = (
  prevDate: IDateContextValues,
  standard: string,
  str: string
) => {
  const { selectedDate } = prevDate[refer(standard)];
  const selectedYear = str.split("-")[0];
  const selectedMonth = str.split("-")[1];

  let newSelectedDate = {
    year: Number(selectedYear),
    month: Number(selectedMonth),
    selectedDate: str,
  };

  if (!str) {
    return {
      ...prevDate,
      [standard]: { ...prevDate[standard], selectedDate: "" },
    };
  }

  if (
    standard !== EStandard.SINGLE &&
    ((refer(standard) === EStandard.ENDDATE && str > selectedDate) ||
      (refer(standard) === EStandard.STARTDATE && str < selectedDate))
  ) {
    return {
      ...prevDate,
      [standard]: { ...prevDate[standard], ...newSelectedDate },
      [refer(standard)]: { ...prevDate[refer(standard)], selectedDate: "" },
    };
  }

  return { ...prevDate, [standard]: { ...prevDate[standard], ...newSelectedDate } };
};

export const onChangeMonth = (prevDate: IDate, index: number) => {
  return { ...prevDate, unit: EUnit.DAY, month: index + 1 };
};

export const onChangeYearOrDecade = (prevDate: IDate, year: number) => {
  return { ...prevDate, unit: prevDate.unit === EUnit.DECADE ? EUnit.YEAR : EUnit.MONTH, year };
};

export const setSelectDate = (prevDate: IDateContextValues, double: boolean, value: string[]) => {
  if (double) {
    return {
      ...prevDate,
      startDate: getDate(prevDate.single, value[0]),
      endDate: getDate(prevDate.single, value[1]),
    };
  } else {
    return {
      ...prevDate,
      single: getDate(prevDate.single, value[0]),
    };
  }
};

const getDate = (prev: any, value: string) => {
  return {
    ...prev,
    selectedDate: value,
    year: getYear(prev, value),
    month: getMonth(prev, value),
  };
};

const getYear = (prev: any, value: string) => {
  if (!value) return prev.year;

  return Number(value.split("-")[0]);
};

const getMonth = (prev: any, value: string) => {
  if (!value) return prev.year;
  return Number(value.split("-")[1]);
};
