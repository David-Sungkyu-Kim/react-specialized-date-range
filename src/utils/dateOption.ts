import { EStandard } from "../@types/date";

export const refer = (standard: string) =>
  standard === EStandard.SINGLE
    ? EStandard.SINGLE
    : standard === EStandard.STARTDATE
    ? EStandard.ENDDATE
    : EStandard.STARTDATE;
