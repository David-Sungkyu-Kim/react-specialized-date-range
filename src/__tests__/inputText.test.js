import "@testing-library/jest-dom";
import React from "react";
import { act, fireEvent, render, renderHook, screen } from "@testing-library/react";
import { useDateContext } from "../hooks/useDateContext";

import DateInput from "../components/DateInput";
import DateProvider from "../context/DateProvider";
import { formattingNumToDate } from "../utils/dateFormat";

import { EStandard, EType } from "../@types/date";

describe("DateInput input value test", () => {
  it("has a input box", () => {
    const { getByRole } = render(<DateInput />);
    getByRole("input");
  });

  it("matches snapshot", () => {
    const { container } = render(<DateInput />);
    expect(container).toMatchSnapshot();
  });

  it("if custom value set input value and dateContext value", () => {
    const customValue = "2022-01-05";

    const { getByRole } = render(<DateInput value={customValue} />);
    const { result } = renderHook(() => useDateContext(), {
      wrapper: ({ children }) => <DateProvider>{children}</DateProvider>,
    });

    act(() => {
      result.current.action.changeHighlightDate(
        EStandard.STARTDATE,
        customValue,
        "YYYY-MM-DD",
        EType.INPUT
      );
    });

    expect(getByRole("input")).toHaveAttribute("value", customValue);
    expect(result.current.value.startDate.selectedDate).toBe(customValue);
  });

  it("should change a value when write in input box", () => {
    const { getByRole } = render(<DateInput />);
    const input = getByRole("input");

    fireEvent.change(input, { target: { value: "20220101" } });

    expect(formattingNumToDate("20220101", "YYYY-MM-DD")).not.toBe("20220101");
    expect(formattingNumToDate("20220101", "YYYY-MM-DD")).toBe("2022-01-01");
    expect(input).toHaveAttribute("value", "2022-01-01");
  });
});
