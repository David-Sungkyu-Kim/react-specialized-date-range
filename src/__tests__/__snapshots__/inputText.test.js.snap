// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput input value test matches snapshot 1`] = `
<div>
  <div
    class="inputWrapper"
  >
    <label
      class="inputLabel"
      role="label"
    >
      end date:
    </label>
    <div
      class="inputBasicWrapper"
    >
      <input
        class="inputValue"
        placeholder=""
        role="input"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
